src/sparql.rs

```diff
@@ -42,13 +44,35 @@ fn term_to_hdt_bgp_str(term: Term) -> String {
     }
 }
 
+// TODO: save space using 32 bit
+type DictIds = [Id; 3];
+
+/// Internal term representation that can be either:
+/// - An ID-based term found in the HDT dictionary (optimized path)
+/// - A computed term not in the dictionary (fallback to Term)
+#[derive(Clone, Eq, PartialEq, Hash, Debug)]
+pub enum InternalTerm {
+    /// Term found in dictionary: [subject_id, predicate_id, object_id]
+    /// Non-zero values indicate valid IDs for that position
+    DictIds([Id; 3]),
+    /// Computed term not in dictionary (e.g., CONCAT results)
+    Computed(Term),
+}
+
+fn term_ids(hdt: &Hdt, id: Id, kind: IdKind) -> DictIds {
+    // is there a more efficient way than going around strings again?
```

Comment:
This is a big point. If you load and allocate strings here, it's likely to drop the main source of optimization of this change.